drivers: i2s(sam ssc): add sys/ring_buffer compatibility wrapper

noobninjanerd opened the pull request:
# drivers: i2s(sam ssc): add `sys/ring_buffer` compatibility wrapper

### References: #21155,  attempts to resolve a part of it.

The `i2s_sam_ssc.c` driver was using a local ring buffer implementation, despite there being a common zephyr `sys/ring_buffer` implementation. This led to redundancy and naming collisions.

I have written wrapper functions around the `sys/ring_buffer`, and added them to this driver. This way, the existing producer/consumer call sites remain preserved, so the rest of the driver doesn't need to change.

This change will simplify future maintenance. 

## Testing
- I have added a small test that checks these wrapper functions with `native_sim`. Running the binary gave the expected results, 
```
$ west build -b native_sim tests/lib/ringbuf_compat_test -p always
$ ./build/zephyr/zephyr.exe
``` 
- I was unable to run the tests using `twister` however. I have probably messed up setting up the test project properly? I have added the terminal log at the bottom.
- Was able to build `samples/drivers/i2s/i2s_api` for several SAM boards in the list. I didn't check all, I am not sure whether it helps. Maybe @nandojve can build and flash to the SAM board they have and verify,
```
$ west build -b sama7g54_ek tests/drivers/i2s/i2s_api -p always
```

## Notes
- I applied `clang-format` to the edited files, but `scripts/checkpatch.pl` still reported a few style issues, which I fixed. 
- I also ran `check_compliance.py` and removed the errors pertaining to my code changes

### Twister log:
```
~/.../zephyr$ scripts/twister -p native_sim -T tests/lib/ringbuf_compat_test -v
ZEPHYR_BASE unset, using "/home/ctrlaarav/zephyr_project/zephyr"
INFO    - Using Ninja..
INFO    - Zephyr version: v4.3.0-rc2-130-g469ee6dd02b6
INFO    - Using 'gnuarmemb' toolchain.
ERROR   - Twister call stack dump:
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/twister", line 208, in <module>
    sys.exit(main())
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 265, in main
    return twister(options, default_options)
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 41, in _inner
    return func(*args, **kwargs)
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 117, in twister
    tplan.discover()
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/testplan.py", line 220, in discover
    raise TwisterRuntimeError("No testsuites found at the specified location...")

Traceback (most recent call last):
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/twister", line 208, in <module>
    sys.exit(main())
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 265, in main
    return twister(options, default_options)
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 41, in _inner
    return func(*args, **kwargs)
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/twister_main.py", line 117, in twister
    tplan.discover()
  File "/home/ctrlaarav/zephyr_project/zephyr/scripts/pylib/twister/twisterlib/testplan.py", line 220, in discover
    raise TwisterRuntimeError("No testsuites found at the specified location...")
twisterlib.error.TwisterRuntimeError: No testsuites found at the specified location...
```

pdgendt commented:
> The stray `CMakeStyles.txt` was not letting some of the checks pass (related to spdx-headers, and text formatting) so I just removed it and did a `git commit --amend`. It gets auto-generated with each build somehow.

But the file is still in the PR, did you forget to force push?

noobninjanerd commented:
I have missed adding the `SPDX-FileCopyrightText: Copyright The Zephyr Project Contributors` header text to `tests/lib/ringbuf_compat_test/src/main.c` and it's showing up in the compliance checks. I will add that and modify the PR. 

Should I wait for the twister tests to finish or do it right away?

noobninjanerd commented:
Removed the trailing whitespaces and ran `check_compliance.py` locally. No more errors related to the relevant files showed up so I am hoping it will the pass compliance checks this time :crossed_fingers:

noobninjanerd commented:
So the current errors in the `Compliance Checks`, only show up as warnings when I run `check_compliance.py` locally. 

Also I didn't make any changes to the lines which the warnings/errors are referencing.

Should I edit the function prototypes in these lines, and give their respective arguments some appropriate names?

pdgendt commented on drivers/i2s/i2s_sam_ssc.c:
```diff
@@ -631,13 +620,11 @@ static int i2s_sam_configure(const struct device *dev, enum i2s_dir dir,
 
-static int rx_stream_start(struct stream *stream, Ssc *const ssc,
-			   const struct device *dev_dma)
+static int rx_stream_start(struct stream *stream, Ssc *const ssc, const struct device *dev_dma)
```
Argument name can't be the same as the struct name

```suggestion
static int rx_stream_start(struct stream *str, Ssc *const ssc, const struct device *dev_dma)
```

noobninjanerd committed:
drivers: i2s(sam ssc): add sys/ring_buffer compatibility wrapper

The driver was implementing a local ring buffer despite there being a
common zephyr sys/ring_buffer implemenation, leading to redundancy and
naming collisions.

I have added wrapper functions that call sys/ring_buffer. This way, no
references to the producer/consumer system in the rest of the
implementation need to be changed.

I have also added a test that checks these wrapper functions and I was
able to verify it using 'west build' for native_sim.

Signed-off-by: Aarav Vaid <vaid.aarav@outlook.com>

noobninjanerd committed:
drivers: i2s(sam ssc): fix identifiers

The function definitions in 'struct stream{}' had arguements that were
missing identifiers (names). Some other function prototypes were also
lacking arguement identifiers. I modified these declarations to give
these arguements identifiers.

The construct 'stream' was also incorrectly declared with the identifier
"stream" (same as the struct type name), in a number of function
definitions. I changed all these identifiers to "str".

Signed-off-by: Aarav Vaid <vaid.aarav@outlook.com>

nandojve commented:
Hi @noobninjanerd ,

> Was able to build samples/drivers/i2s/i2s_api for several SAM boards in the list. I didn't check all, I am not sure whether it helps. Maybe @nandojve can build and flash to the SAM board they have and verify,

Which boards did you built and ran ?

noobninjanerd commented:
> Which boards did you built and ran ?

Hi, I only built the binaries, I didn't flash and run any since I don't have the boards in question. And they were just the first 3-4 boards in the SAM boards list.

I can procure one and flash it, do you have any (preferably cheap) recommendations?

noobninjanerd commented:
Hi @nandojve, thank you for trying out the code changes! 

1. `west build -b native_sim tests/lib/ringbuf_compat_test/ -p always` does build for me, albeit with a few warnings:
```
(.venv) ctrlaarav@Aarav-Lab:~/.../zephyr$ west build -b native_sim tests/lib/ringbuf_compat_test/ -p always
-- west build: making build dir /home/ctrlaarav/zephyr_project/zephyr/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/ctrlaarav/zephyr_project/zephyr/tests/lib/ringbuf_compat_test
-- CMake version: 3.22.1
-- Found Python3: /home/ctrlaarav/zephyr_project/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.10") found components: Interpreter 
-- Cache files will be written to: /home/ctrlaarav/.cache/zephyr
-- Zephyr version: 4.3.0-rc2 (/home/ctrlaarav/zephyr_project/zephyr)
-- Found west (found suitable version "1.5.0", minimum required is "0.14.0")
-- Board: native_sim, qualifiers: native
-- Found toolchain: host (gcc/ld)
-- Found Dtc: /usr/bin/dtc (found suitable version "1.6.1", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/ctrlaarav/zephyr_project/zephyr/boards/native/native_sim/native_sim.dts
-- Generated zephyr.dts: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/zephyr.dts
-- Generated pickled edt: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/edt.pickle
-- Generated devicetree_generated.h: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/include/generated/zephyr/devicetree_generated.h

warning: STDOUT_CONSOLE (defined at lib/libc/Kconfig:169) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: (!(NATIVE_LIBRARY && EXTERNAL_LIBC)) (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_STDOUT_CONSOLE and/or look up
STDOUT_CONSOLE in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

Parsing /home/ctrlaarav/zephyr_project/zephyr/Kconfig
Loaded configuration '/home/ctrlaarav/zephyr_project/zephyr/boards/native/native_sim/native_sim_defconfig'
Merged configuration '/home/ctrlaarav/zephyr_project/zephyr/tests/lib/ringbuf_compat_test/prj.conf'
Configuration saved to '/home/ctrlaarav/zephyr_project/zephyr/build/zephyr/.config'
Kconfig header saved to '/home/ctrlaarav/zephyr_project/zephyr/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /usr/bin/ld.bfd (found version "2.38") 
-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/gcc
-- Using ccache: /usr/bin/ccache
-- Found gen_kobject_list: /home/ctrlaarav/zephyr_project/zephyr/scripts/build/gen_kobject_list.py
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ctrlaarav/zephyr_project/zephyr/build
-- west build: building application
[1/104] Preparing syscall dependency handling

[2/104] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.3.0-rc2 (/home/ctrlaarav/zephyr_project/zephyr), build: v4.3.0-rc2-131-g1fea38e6efc6
[98/104] Building C object CMakeFiles/app.dir/src/main.c.obj
In file included from /home/ctrlaarav/zephyr_project/zephyr/subsys/testsuite/ztest/include/zephyr/ztest.h:44,
                 from /home/ctrlaarav/zephyr_project/zephyr/tests/lib/ringbuf_compat_test/src/main.c:7:
/home/ctrlaarav/zephyr_project/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_test.h:238:26: warning: initialization of ‘void * (*)(void)’ from incompatible pointer type ‘void (*)(void)’ [-Wincompatible-pointer-types]
  238 |                 .setup = (setup_fn),                                                               \
      |                          ^
/home/ctrlaarav/zephyr_project/zephyr/tests/lib/ringbuf_compat_test/src/main.c:84:1: note: in expansion of macro ‘ZTEST_SUITE’
   84 | ZTEST_SUITE(ringbuf_tests, NULL, tb_init, NULL, NULL, NULL);
      | ^~~~~~~~~~~
/home/ctrlaarav/zephyr_project/zephyr/subsys/testsuite/ztest/include/zephyr/ztest_test.h:238:26: note: (near initialization for ‘z_ztest_test_node_ringbuf_tests.setup’)
  238 |                 .setup = (setup_fn),                                                               \
      |                          ^
/home/ctrlaarav/zephyr_project/zephyr/tests/lib/ringbuf_compat_test/src/main.c:84:1: note: in expansion of macro ‘ZTEST_SUITE’
   84 | ZTEST_SUITE(ringbuf_tests, NULL, tb_init, NULL, NULL, NULL);
      | ^~~~~~~~~~~
[102/104] Linking C executable zephyr/zephyr.elf
Generating files from /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/zephyr.elf for board: native_sim
[104/104] Running utility command for native_runner_executable
(.venv) ctrlaarav@Aarav-Lab:~/.../zephyr$ ./build/zephyr/zephyr.exe 
*** Booting Zephyr OS build v4.3.0-rc2-131-g1fea38e6efc6 ***
Running TESTSUITE ringbuf_tests
===================================================================
START - test_put_get
 PASS - test_put_get in 0.000 seconds
===================================================================
TESTSUITE ringbuf_tests succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [ringbuf_tests]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.000 seconds
 - PASS - [ringbuf_tests.test_put_get] duration = 0.000 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```
You can see that the test also runs. However, I will address the incorrect return type warning for the setup function (`tb_init()`) and make the appropriate changes.

2. I guess it would have been better if I did build for all the boards :sweat_smile:, these build errors are new. I was able to reproduce the errors on my system too, and I am able to fix them too,
 i. Since `buf_size` is not a compile-time constant (which is what `BUILD_ASSERT` expects), I will replace `BUILD_ASSERT` with `__ASSERT`  (which is a  runtime check)
 ii. `rx_0_ring_bug` and `tx_0_ring_buf` have been defined in the file, but after the functions which are making the calls to them. I have simply shifted their declarations to near the beginning of the file.
 
And it builds!
 ```
~/.../zephyr$ west build -b sam_v71_xult/samv71q21 tests/drivers/i2s/i2s_api/ -p always
-- west build: making build dir /home/ctrlaarav/zephyr_project/zephyr/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /home/ctrlaarav/zephyr_project/zephyr/tests/drivers/i2s/i2s_api
-- CMake version: 3.22.1
-- Found Python3: /home/ctrlaarav/zephyr_project/.venv/bin/python3 (found suitable version "3.10.12", minimum required is "3.10") found components: Interpreter 
-- Cache files will be written to: /home/ctrlaarav/.cache/zephyr
-- Zephyr version: 4.3.0-rc2 (/home/ctrlaarav/zephyr_project/zephyr)
-- Found west (found suitable version "1.5.0", minimum required is "0.14.0")
-- Board: sam_v71_xult, qualifiers: samv71q21
-- Found toolchain: gnuarmemb (/usr)
-- Found Dtc: /usr/bin/dtc (found suitable version "1.6.1", minimum required is "1.4.6") 
-- Found BOARD.dts: /home/ctrlaarav/zephyr_project/zephyr/boards/atmel/sam/sam_v71_xult/sam_v71_xult_samv71q21.dts
-- Found devicetree overlay: /home/ctrlaarav/zephyr_project/zephyr/tests/drivers/i2s/i2s_api/boards/sam_v71_xult_samv71q21.overlay
-- Generated zephyr.dts: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/zephyr.dts
-- Generated pickled edt: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/edt.pickle
-- Generated devicetree_generated.h: /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/include/generated/zephyr/devicetree_generated.h
Parsing /home/ctrlaarav/zephyr_project/zephyr/tests/drivers/i2s/i2s_api/Kconfig
Loaded configuration '/home/ctrlaarav/zephyr_project/zephyr/boards/atmel/sam/sam_v71_xult/sam_v71_xult_samv71q21_defconfig'
Merged configuration '/home/ctrlaarav/zephyr_project/zephyr/tests/drivers/i2s/i2s_api/prj.conf'
Configuration saved to '/home/ctrlaarav/zephyr_project/zephyr/build/zephyr/.config'
Kconfig header saved to '/home/ctrlaarav/zephyr_project/zephyr/build/zephyr/include/generated/zephyr/autoconf.h'
-- Found GnuLd: /usr/share/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/arm-none-eabi/bin/ld.bfd (found version "2.43.1") 
-- The C compiler identification is GNU 14.2.1
-- The CXX compiler identification is GNU 14.2.1
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/arm-none-eabi-gcc
-- Using ccache: /usr/bin/ccache
-- Found gen_kobject_list: /home/ctrlaarav/zephyr_project/zephyr/scripts/build/gen_kobject_list.py
-- Configuring done
-- Generating done
-- Build files have been written to: /home/ctrlaarav/zephyr_project/zephyr/build
-- west build: building application
[9/1158] Preparing syscall dependency handling

[27/1158] Generating include/generated/zephyr/version.h
-- Zephyr version: 4.3.0-rc2 (/home/ctrlaarav/zephyr_project/zephyr), build: v4.3.0-rc2-131-g1fea38e6efc6
[1139/1158] Linking C executable zephyr/zephyr_pre0.elf
/usr/share/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/14.2.1/../../../../arm-none-eabi/bin/ld.bfd: warning: zephyr/zephyr_pre0.elf has a LOAD segment with RWX permissions
[1148/1158] Linking C executable zephyr/zephyr_pre1.elf
/usr/share/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/14.2.1/../../../../arm-none-eabi/bin/ld.bfd: warning: zephyr/zephyr_pre1.elf has a LOAD segment with RWX permissions
[1158/1158] Linking C executable zephyr/zephyr.elf
/usr/share/arm-gnu-toolchain-14.2.rel1-x86_64-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/14.2.1/../../../../arm-none-eabi/bin/ld.bfd: warning: zephyr/zephyr.elf has a LOAD segment with RWX permissions
Memory region         Used Size  Region Size  %age Used
           FLASH:      102004 B         2 MB      4.86%
             RAM:       15488 B       384 KB      3.94%
        IDT_LIST:           0 B        32 KB      0.00%
Generating files from /home/ctrlaarav/zephyr_project/zephyr/build/zephyr/zephyr.elf for board: sam_v71_xult
 ```

noobninjanerd commented on drivers/i2s/i2s_sam_ssc.c:
```diff
@@ -100,70 +97,88 @@ struct i2s_sam_dev_data {
+static void ring_buf_compat_init(struct ring_buf *rb, uint8_t *buf, size_t buf_size)
 {
-	unsigned int key;
+	BUILD_ASSERT((buf_size % ENTRY_SIZE) == 0,
```
```suggestion
	__ASSERT((buf_size % ENTRY_SIZE) == 0,
```
